x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/queries/WeightedTokensQueryBuilder.java

```diff
@@ -0,0 +1,247 @@
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.startObject(fieldName);
+        builder.startObject(TOKENS_FIELD.getPreferredName());
+        for (var token : tokens) {
+            token.toXContent(builder, params);
+        }
+        builder.endObject();
+        if (threshold != null) {
+            threshold.toXContent(builder, params);
+        }
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+        builder.endObject();
+    }
+
+    private float getAverageTokenFreqRatio(IndexReader reader, int fieldDocCount) throws IOException {
+        int numUniqueTokens = 0;
+        for (var leaf : reader.getContext().leaves()) {
+            var terms = leaf.reader().terms(fieldName);
+            if (terms != null) {
+                numUniqueTokens = (int) Math.max(terms.size(), numUniqueTokens);
+            }
+        }
+        if (numUniqueTokens == 0) {
+            return 0;
+        }
+        return (float) reader.getSumDocFreq(fieldName) / fieldDocCount / numUniqueTokens;
+    }
+
+    /**
+     * Returns true if the token should be queried based on the {@code ratioThreshold} and {@code weightThreshold}
+     * set on the query.
+     */
+    private boolean shouldKeepToken(
+        IndexReader reader,
+        WeightedToken token,
+        int fieldDocCount,
+        float averageTokenFreqRatio,
+        float bestWeight
+    ) throws IOException {
+        if (threshold == null) {
+            return true;
+        }
+        int docFreq = reader.docFreq(new Term(fieldName, token.token()));
+        if (docFreq == 0) {
+            return false;
+        }
+        float tokenFreqRatio = (float) docFreq / fieldDocCount;
+        return tokenFreqRatio < threshold.getRatioThreshold() * averageTokenFreqRatio
+            || token.weight() > threshold.getWeightThreshold() * bestWeight;
+    }
+
+    @Override
+    protected Query doToQuery(SearchExecutionContext context) throws IOException {
+        final MappedFieldType ft = context.getFieldType(fieldName);
+        if (ft == null) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against a field that does not exist");
+        }
+        var qb = new BooleanQuery.Builder();
+        int fieldDocCount = context.getIndexReader().getDocCount(fieldName);
+        float bestWeight = 0f;
+        for (var t : tokens) {
+            bestWeight = Math.max(t.weight(), bestWeight);
+        }
+        float averageTokenFreqRatio = getAverageTokenFreqRatio(context.getIndexReader(), fieldDocCount);
+        if (averageTokenFreqRatio == 0) {
+            return new MatchNoDocsQuery("The \"" + getName() + "\" query is against an empty field");
+        }
+        for (var token : tokens) {
+            boolean keep = shouldKeepToken(context.getIndexReader(), token, fieldDocCount, averageTokenFreqRatio, bestWeight) ^ threshold
```

Comment:
I think we're OK here, because the constructor has been updated to: 
```
this.threshold = Objects.requireNonNullElse(threshold, new WeightedTokensThreshold());
```

So if no `threshold` is specified we'll go with the defaults.

Reply:
That's reasonable, I can update.